docs/design/bucket.md

```diff
@@ -0,0 +1,103 @@
+
+When high performance workload running on Kubernetes, the performance is greatly affected by pod affinity/anti-affinity, e.g.
+ps/worker mode of Tensorflow. The default scheduler support pod affinity/anti-affinity, but there're some weaknesses:
+
+- The default scheduler will check PodAffinity/PodAntiAffinity against all running pods which impacts schedule performance/throughput
+- The default scheduler only check current pod, the coming pods may not have enough resources to meet affinity
+
+To improve the performance of scheduler and workload, a new algorithm is introduced, named as Bucket. This doc describes the function
+specification of the new algorithm.
+
+## Function Specification
+
+### Scope
+
+* In Scope
+
+  - Provide a way for user to define task affinity/anti-affinity within a job
+  - Support selecting a better node for the tasks based on affinity/anti-affinity
+
+* Out Of Scope
+
+  - Do not support "Network Topology", e.g. switches  
+  
+### API
+
+A new plugin is introduced to avoid Job API changes, named as `task-topology`. The arguments of the plugin are defined
+as follow: 
+
+```yaml
+plugins
+  task-topology: ["--affinity", "[[ps-task, worker-task]]", "--anti-affinity", "[[ps-task]]"]
+```
+
+**affinity:**
+
+The `--affinity` parameter defines which tasks/pods should be placed on the same node. It's a json two-dimension array:
+
+- First dimension defines which tasks affinity, e.g. `[ps-task, worker-task]`; it only defines the affinity between tasks.
+  The array with only one tasks defines affinity to itself.  
+- Second dimension defines multiple affinity configuration, e.g. `[[ps, worker], [worker]]` means `ps` and `worker` affinity
+  to each other, and `worker` affinity to itself, but not self-affinity preference for `ps` 
+
+**anti-affinity:**
+
+The `--anti-affinity` parameter defines which tasks/pods should not be placed on the same node. It's also a json two-dimension array:
+
+- First dimension defines which tasks anti-affinity, e.g. `[ps-task, worker-task]`; it only defines the anti-affinity between tasks.
+  The array with only one tasks defines anti-affinity to itself.  
+- Second dimension defines multiple anti-affinity configuration, e.g. `[[ps, worker], [ps]]` means `ps` and `worker` anti-affinity
+  to each other, and `ps` anti-affinity to itself, but not self-affinity preference for `worker` 
+
+If there's any conflict, the pod will not be placed in any bucket.
+
+### Function Details
+
+#### Job controller
+
+In `OnJobAdd` callback, the `task-topology` plugin will encode the arguments of `task-topology` into a json string and
+put them into `Job`'s annotation if `volcano.sh/task-topology` not found.
+
+When job controller create `PodGroup` for the job, all its annotations and labels will be passed on to the `PodGroup`. 
+
+#### Bucket Plugin
+
+##### OnSessionOpen
+
+In `OnSessionOpen` of `bucket` plugin, it'll get `task-topology` from `PodGroup`'s annotation and build bucket accordingly.
```

Comment:
I think we have to consider --affinity and --anti-affinity  differently  while choosing the node for a tasks related to job.  as bucket is group of  related tasks of a job.